### PR TITLE
Add support for Python 3.12

### PR DESCRIPTION
**Pull request by tkumor3**
Closes #

## Introduced changes

- 
- 

##

- [ ] This PR contains breaking changes

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -39,18 +39,17 @@ poethepoet = "^0.24.4"
...
 isort = "^5.11.4"
 pytest-rerunfailures = "^13.0"
 python-dotenv = "^1.0.0"
+setuptools = "^69.0.3"
 
-
-[tool.poetry.group.py39-dev.dependencies]
-cairo-lang = {version = "0.13.0", python = ">=3.9, <3.10"}
+[tool.poetry.group.py312-dev.dependencies]
+cairo-lang = {version = "0.13.0", python = ">=3.9, <3.13"}
```
Comment (ddoktorski) [inline pyproject.toml:52]: Please shift this to the `dev.dependencies` section, omit the Python requirement, and accordingly adjust the workflow.

**Review by ddoktorski — COMMENTED** [review]
I've left one comment, but overall it looks good 💪 

Please also revise the [Development dependencies](https://starknetpy.readthedocs.io/en/latest/development.html#development-dependencies) section in the docs to remove the requirement for Python 3.9.x

File: pyproject.toml (hunk 1 above)
Comment (tkumor3) [reply pyproject.toml]: It doesnt works :( Maybe I did sth wrong could you take a look?

File: pyproject.toml [hunk 2]
```diff
@@ -14,10 +14,10 @@ repository = "https://github.com/software-mansion/starknet.py"
 documentation = "https://starknetpy.rtfd.io/"
 
 [tool.poetry.dependencies]
-python = ">=3.8, <3.12"
+python = ">=3.9, <3.13"
 asgiref = "^3.4.1"
 marshmallow = "^3.15.0"
-marshmallow-oneofschema = "^3.0.1"
+marshmallow-oneofschema = "3.0.1"
```
Comment (ddoktorski) [inline pyproject.toml:20]: Why this change?

**Review by ddoktorski — CHANGES_REQUESTED** [review]

File: .github/workflows/checks.yml [hunk 3]
```diff
@@ -110,7 +112,6 @@ jobs:
           key: ${{ runner.os }}-contracts-${{ hashFiles('starknet_py/tests/e2e/mock/contracts', 'poetry.lock') }}
 
       - name: Compile contracts
-        if: steps.cache-contracts.outputs.cache-hit != 'true'
```
[the remark this answers is not in the document]
Comment (tkumor3) [reply .github/workflows/checks.yml:113]: It should be reverted! it was added by testing purpose

File: pyproject.toml (hunk 2 above)
Comment (tkumor3) [reply pyproject.toml:20]: update to 3.1.0 create some pyright issues

File: .github/workflows/checks.yml [hunk 4]
```diff
@@ -238,20 +238,19 @@ jobs:
           pip install poetry
           poetry config installer.modern-installation false
 
-      - name: Install python 3.9 requirements
+      - name: Install python 3.12 requirements
         run: |
-          poetry export -f requirements.txt --only=py39-dev --without-hashes --output requirements.txt
-          pip install -r requirements.txt
+          pip install cairo-lang==0.13.0
```
[the remark this answers is not in the document]
Comment (tkumor3) [reply .github/workflows/checks.yml:243]: in actions bellow we use poetry
```      - name: Set up Python ${{ matrix.python-version }}
        uses: actions/setup-python@v4
        with:
          python-version: ${{ matrix.python-version }}
          cache: 'poetry'

      - name: Install dependencies
        run: |
          poetry install
```

File: .github/workflows/checks.yml [hunk 5]
```diff
@@ -98,9 +97,12 @@ jobs:
 
       - name: Install dependencies
         run: |
-          poetry install --without py39-dev
+          poetry install
 
       # ====================== CONTRACTS v0 ====================== #
+      - name: Install setuptools
+        run: |
+          pip install --upgrade setuptools
```
[the remark this answers is not in the document]
Comment (tkumor3) [reply .github/workflows/checks.yml:105]: python 3.12 doesn't install it in the default

File: .github/workflows/checks.yml [hunk 6]
```diff
@@ -131,7 +133,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python-version: [ "3.8", "3.9", "3.10", "3.11" ]
+        python-version: ["3.8", "3.9", "3.10", "3.11", "3.12" ]
```
Comment (ddoktorski) [inline .github/workflows/checks.yml:136]: nit:
```suggestion
        python-version: [ "3.8", "3.9", "3.10", "3.11", "3.12" ]
```

File: .github/workflows/manual_tests.yml [hunk 7]
```diff
@@ -13,10 +13,10 @@ on:
         type: choice
         description: 'Python version'
         options:
-        - '3.8'
```
Comment (ddoktorski) [inline .github/workflows/manual_tests.yml:16]: I guess we don't use this workflow at all, so we might think about getting rid of it. But what is the reason for removing `3.8` now?

File: docs/development.rst [hunk 8]
```diff
@@ -9,8 +9,7 @@ Development dependencies
 ------------------------
 - `poetry <https://python-poetry.org/>`_ - dependency manager.
 - `pyenv <https://github.com/pyenv/pyenv>`_ - recommended for installing and switching python versions locally.
-
-Make sure running ``poetry run python --version`` returns ``Python 3.9.x``.
+- `cairo-lang <https://pypi.org/project/cairo-lang/>`_ - required to compail contract  (`poe compile_contracts`)
```
Comment (ddoktorski) [inline docs/development.rst:12]:
```suggestion
- `cairo-lang <https://pypi.org/project/cairo-lang/>`_ - required to compile contracts  (`poe compile_contracts`)
```

File: .github/workflows/package.yml [hunk 9]
```diff
@@ -9,7 +9,7 @@ jobs:
       - uses: actions/checkout@v3
       - uses: actions/setup-python@v4
         with:
-          python-version: '3.9'
+          python-version: '3.12'
```
Comment (ddoktorski) [inline .github/workflows/package.yml:12]: I would stick with `3.9` since I don't see any advantage in updating it at this point.

File: .github/workflows/checks.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (tkumor3) [reply .github/workflows/checks.yml:105]: Yep, that was the first solution! We decided with @THenry14 to use 3.12 everywhere where possible.

[the remark this answers is not in the document]
Comment (ddoktorski) [reply .github/workflows/checks.yml:105]: I guess if we use `3.11` here we can also get rid of `setuptools` from `pyproject.toml`, is that correct?

[the remark this answers is not in the document]
Comment (tkumor3) [reply .github/workflows/checks.yml]: I can check it

[the remark this answers is not in the document]
Comment (tkumor3) [reply .github/workflows/checks.yml]: it works

File: .github/workflows/checks.yml [hunk 10]
```diff
@@ -282,7 +272,7 @@ jobs:
...
```
Comment (ddoktorski) [inline .github/workflows/checks.yml:275]: nit:
```suggestion
        python-version: [ "3.8", "3.9", "3.10", "3.11", "3.12" ]
```
It can be found also in a few other places.

**Review by ddoktorski — APPROVED** [review]
I've left one more minor comment. 

Please rename the title before merging 👍
